### PR TITLE
Problem: GetEthBalance blocks the GameThread (fix #270)

### LINKED ISSUE #270: Problem: GetEthBalance blocks the GameThread
If can not get the balance (Timeout or whatever), the blocking may take 10+ seconds.

### PR DESCRIPTION
**Pull request by leejw51crypto**
Solution: convert sync apis to async

change GetEthBalance to async

change GetBalance to async

convert GetNFTSupply to async

convert GetNFTOwner to async

convert GetNFTCollection to async

convert GetNFTDenom to async

change GetNFTDenomByName to async

convert GetNFTAllDenoms to async

refactor sendamount

make SendAmount to async

### CODE REVIEW

**Comment by leejw51crypto** [comment]
coverting more apis to async

**Review by tomtau — COMMENTED** [review]
the change on Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp is quite big -- I guesss it changes other functions to be async as well?

**Comment by leejw51crypto** [comment]
yes, changed all query apis to async

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (tomtau) [inline Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:370]: is this cast safe? should there be an assertion that amount is non-negative?

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (tomtau) [inline Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:1321]: why does `.legacy()` need to be called?

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (tomtau) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:33]: why does it need these delegate declarations for all different types (bool, int, string...)?

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (tomtau) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:42]: the name says `FGeneralCronosCallInt32Delegate`, but the argument seems to be int64 -- not sure if intended, or duplicate (enough to have `FGeneralCronosCallInt64Delegate`?) ...

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (tomtau) [reply Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:33]: oh because these general ones are used over it in different functions?

Comment (tomtau) [reply Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:33]: maybe good to document it

**Review by tomtau — APPROVED** [review]
DefiWalletCoreActor.cpp's change is too large, so I only had a quick skim through it -- I assume the code was mainly just copied over / wrapped around the async boilerplate, so it should be as before, but not sure if any issues sneaked it.

From the quick look, I noticed a few minor issues, but they were already in the original code, so it could perhaps be fixed in a different PR.

One alternative that would be easier to review would be to split this PR to be function-by-function, but that may be a bit tedious. overall, probably ok to merge this, maybe @damoncro can make the final call

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (tomtau) [reply Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:1321]: oh I see it in many places, but it was in the original code

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:356]: Same as other comments, better off assigning it a dedicated delegate.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:452]: Same, better off assigning a dedicated delegate.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:544]: Same, better off assigning a dedicated delegate.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:556]: Same, better off assigning a dedicated delegate.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:570]: Same, better off assigning a dedicated delegate.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:596]: Same, better off assigning a dedicated delegate.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:586]: Same, better off assigning a dedicated delegate.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:606]: Same, better off assigning a dedicated delegate.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:556]: Same, better off assigning a dedicated delegate.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:617]: Same, better off assigning a dedicated delegate. Rename it as well.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:628]: Same, better off assigning a dedicated delegate. Rename it as well.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:638]: Same, better off assigning a dedicated delegate. Rename it as well.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:648]: Same, better off assigning a dedicated delegate. Rename it as well.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:660]: Same, better off assigning a dedicated delegate. Rename it.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:672]: Same, better off assigning a dedicated delegate. Rename it.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:687]: Same, better off assigning a dedicated delegate. Rename it.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:700]: Same, better off assigning a dedicated delegate. Rename it.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:711]: Same, better off assigning a dedicated delegate. Rename it.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:724]: Same, better off assigning a dedicated delegate. Rename it.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:740]: Same, better off assigning a dedicated delegate. Rename it.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:752]: Same, better off assigning a dedicated delegate. Rename it.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:767]: Same, better off assigning a dedicated delegate. Rename it.

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (damoncro) [inline Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:830]: Same, better off assigning a dedicated delegate. Rename it.

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (damoncro) [inline Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:100]: int64_t or int64?

**Review by damoncro — CHANGES_REQUESTED** [review]
Overall, it is good, ok to merge. However, wrapping AsyncTask to functions are breaking change, better off testing them before merge. I suggest, one PR one function, do microtest/unittest on metaverse (truly use the function on the demo), this also make sure the code quality is high.

We are fixing GetEthBalance here, how about just fix it and make it work on metaverse demo?

**Comment by leejw51crypto** [comment]
working on feedbacks

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:370]: yes, it can raise overflow issue, i'll add exception handling

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:1321]: legacy: not use  EIP-1559
not legacy: use  EIP-1559

default is using  EIP-1559

i'll check with  EIP-1559 off, and remove legacy()

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:33]: i think  , it's better to rename for each delegate.
(each function have each delegate)
I'll change

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:42]: it's intended, but i'll change

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:410]: ok

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:368]: yes, i'll change
each function have each delegate

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:356]: ok

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:452]: ok

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:544]: ok

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:752]: ok

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:767]: ok

File: Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Public/DefiWalletCoreActor.h:830]: ok

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:100]: int64_t is defined in the standard header `<cstdint>`.
so better for compatibility.
i'll change to int64_t

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (tomtau) [reply Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:1321]: 1) legacy isn't needed for Cronos, 2) this flag is turned even on query calls, where no transaction is submitted, so it doesn't make a difference there

**Comment by leejw51crypto** [comment]
continue working

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:1324]: yes, i'll remove .legacy()

**Review by rarcifa — APPROVED** [review]
lgtm

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (damoncro) [reply Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:975]: I saw some functions AsyncTask is outside try...catch.. block, but some are not, better off making them all outside try...catch...

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:100]: ok

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:112]: ok

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:975]: yes, try, catch should be inside AsyncTask.
i'll double check

**Comment by leejw51crypto** [comment]
also i'll change  rust::cxxbridge1::Error to std::exception,
as another pr

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp
Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp]: changed, just one AsyncTask 
no AsyncTask inside catch block

**Review by damoncro — APPROVED** [review]

**Comment by leejw51crypto** [comment]
i'll resolve conflict

**Comment by leejw51crypto** [comment]
submit later

**Pull request closed without merging**
